File: 回測模組/analytics/trade_statistics.py
"""
交易統計計算模組
負責計算交易相關的統計數據
"""
from typing import Dict, List
from strategy_modules.position_manager import TradeRecord


class TradeStatisticsCalculator:
    """交易統計計算器"""

    def __init__(self, data_processor=None):
        """
        初始化統計計算器

        Args:
            data_processor: 資料處理器，用於取得公司名稱等資訊
        """
        self.data_processor = data_processor

    def calculate_statistics(self, stock_id: str, trades: List[TradeRecord], date: str) -> Dict:
        """
        計算交易統計數據

        Args:
            stock_id: 股票代碼
            trades: 交易記錄列表
            date: 交易日期

        Returns:
            統計數據字典
        """
        # 取得公司名稱
        company_name = self.data_processor.get_company_name(stock_id) if self.data_processor else stock_id

        # 基本統計
        total_trades = len(trades)
        win_trades = 0
        loss_trades = 0
        total_entry_price = 0.0
        total_exit_price = 0.0
        total_pnl = 0.0

        for trade in trades:
            # 記錄進場價格
            entry_price = trade.entry_price
            total_entry_price += entry_price

            # 計算出場價格
            exit_price = self._calculate_exit_price(trade)
            total_exit_price += exit_price

            # 計算單筆損益
            single_pnl = (exit_price - entry_price) * 1000  # 每筆交易1張
            total_pnl += single_pnl

            # 判斷勝負
            if exit_price > trade.entry_price:
                win_trades += 1
            elif exit_price < trade.entry_price:
                loss_trades += 1

        # 計算平均價格和勝率
        avg_entry_price = total_entry_price / total_trades if total_trades > 0 else 0
        avg_exit_price = total_exit_price / total_trades if total_trades > 0 else 0
        win_rate = (win_trades / total_trades * 100) if total_trades > 0 else 0

        return {
            'stock_id': stock_id,
            '股票名稱': company_name,
            '進場次數': total_trades,
            '停損次數': loss_trades,
            '勝率(%)': round(win_rate, 2),
            '進場價格': round(avg_entry_price, 2),
            '出場價格': round(avg_exit_price, 2),
            '損益': round(total_pnl, 0)
        }

    def _calculate_exit_price(self, trade: TradeRecord) -> float:
        """
        計算實際出場價格

        Args:
            trade: 交易記錄

        Returns:
            出場價格
        """
        exit_price = 0.0

        if trade.trailing_exit_details:
            # 移動停利出場：計算加權平均價
            total_ratio = 0.0
            weighted_price = 0.0
            for exit in trade.trailing_exit_details:
                weighted_price += exit['price'] * exit['ratio']
                total_ratio += exit['ratio']

            # 加上最終出場（如進場價保護）
            if trade.final_exit_price and total_ratio < 1.0:
                remaining_ratio = 1.0 - total_ratio
                weighted_price += trade.final_exit_price * remaining_ratio
                total_ratio = 1.0

            exit_price = weighted_price / total_ratio if total_ratio > 0 else trade.entry_price
        elif trade.partial_exit_price and trade.final_exit_price:
            # 兩階段出場（原邏輯）
            exit_price = (trade.partial_exit_price * 0.5 + trade.final_exit_price * 0.5)
        elif trade.partial_exit_price:
            exit_price = trade.partial_exit_price
        elif trade.final_exit_price:
            exit_price = trade.final_exit_price
        else:
            exit_price = trade.entry_price

        return exit_price

File: 回測模組/CSharp/BacktestModule/Analytics/TradeStatistics.cs
using System;
using System.Collections.Generic;
using System.Linq;
using BacktestModule.Core;

namespace BacktestModule.Analytics
{
    /// <summary>
    /// Trade statistics calculator.
    /// Computes trading statistics such as win rate, average entry/exit prices, and PnL.
    /// </summary>
    public class TradeStatisticsCalculator : ITradeStatisticsCalculator
    {
        private readonly IDataProcessor _dataProcessor;

        /// <summary>
        /// Initializes the statistics calculator.
        /// </summary>
        /// <param name="dataProcessor">Data processor for retrieving company name information. Can be null.</param>
        public TradeStatisticsCalculator(IDataProcessor dataProcessor = null)
        {
            _dataProcessor = dataProcessor;
        }

        /// <summary>
        /// Calculates trading statistics for a given stock.
        /// </summary>
        /// <param name="stockId">Stock code.</param>
        /// <param name="trades">List of trade records.</param>
        /// <param name="date">Trading date string.</param>
        /// <returns>Dictionary containing statistics with Chinese key names for CSV compatibility.</returns>
        public Dictionary<string, object> CalculateStatistics(string stockId, List<TradeRecord> trades, string date)
        {
            // Get company name
            string companyName = _dataProcessor != null
                ? _dataProcessor.GetCompanyName(stockId)
                : stockId;

            // Basic statistics
            int totalTrades = trades.Count;
            int winTrades = 0;
            int lossTrades = 0;
            double totalEntryPrice = 0.0;
            double totalExitPrice = 0.0;
            double totalPnl = 0.0;

            foreach (var trade in trades)
            {
                // Record entry price
                double entryPrice = trade.EntryPrice;
                totalEntryPrice += entryPrice;

                // Calculate exit price
                double exitPrice = CalculateExitPrice(trade);
                totalExitPrice += exitPrice;

                // Calculate single trade PnL (1 lot = 1000 shares per trade)
                double singlePnl = (exitPrice - entryPrice) * 1000;
                totalPnl += singlePnl;

                // Determine win/loss
                if (exitPrice > trade.EntryPrice)
                {
                    winTrades++;
                }
                else if (exitPrice < trade.EntryPrice)
                {
                    lossTrades++;
                }
            }

            // Calculate averages and win rate
            double avgEntryPrice = totalTrades > 0 ? totalEntryPrice / totalTrades : 0;
            double avgExitPrice = totalTrades > 0 ? totalExitPrice / totalTrades : 0;
            double winRate = totalTrades > 0 ? (double)winTrades / totalTrades * 100 : 0;

            return new Dictionary<string, object>
            {
                ["stock_id"] = stockId,
                ["\u80A1\u7968\u540D\u7A31"] = companyName,                    // 股票名稱
                ["\u9032\u5834\u6B21\u6578"] = totalTrades,                      // 進場次數
                ["\u505C\u640D\u6B21\u6578"] = lossTrades,                       // 停損次數
                ["\u52DD\u7387(%)"] = Math.Round(winRate, 2),                    // 勝率(%)
                ["\u9032\u5834\u50F9\u683C"] = Math.Round(avgEntryPrice, 2),     // 進場價格
                ["\u51FA\u5834\u50F9\u683C"] = Math.Round(avgExitPrice, 2),      // 出場價格
                ["\u640D\u76CA"] = Math.Round(totalPnl, 0)                       // 損益
            };
        }

        /// <summary>
        /// Calculates the actual exit price for a trade, considering trailing stops,
        /// two-stage exits, or single exits.
        /// </summary>
        /// <param name="trade">The trade record.</param>
        /// <returns>The weighted average exit price.</returns>
        public double CalculateExitPrice(TradeRecord trade)
        {
            double exitPrice = 0.0;

            if (trade.TrailingExitDetails != null && trade.TrailingExitDetails.Count > 0)
            {
                // Trailing stop exit: calculate weighted average price
                double totalRatio = 0.0;
                double weightedPrice = 0.0;

                foreach (var exit in trade.TrailingExitDetails)
                {
                    double price = exit.ContainsKey("price") ? Convert.ToDouble(exit["price"]) : 0.0;
                    double ratio = exit.ContainsKey("ratio") ? Convert.ToDouble(exit["ratio"]) : 0.0;
                    weightedPrice += price * ratio;
                    totalRatio += ratio;
                }

                // Add final exit (e.g., entry price protection)
                if (trade.FinalExitPrice.HasValue && totalRatio < 1.0)
                {
                    double remainingRatio = 1.0 - totalRatio;
                    weightedPrice += trade.FinalExitPrice.Value * remainingRatio;
                    totalRatio = 1.0;
                }

                exitPrice = totalRatio > 0 ? weightedPrice / totalRatio : trade.EntryPrice;
            }
            else if (trade.PartialExitPrice.HasValue && trade.FinalExitPrice.HasValue)
            {
                // Two-stage exit (original logic)
                exitPrice = trade.PartialExitPrice.Value * 0.5 + trade.FinalExitPrice.Value * 0.5;
            }
            else if (trade.PartialExitPrice.HasValue)
            {
                exitPrice = trade.PartialExitPrice.Value;
            }
            else if (trade.FinalExitPrice.HasValue)
            {
                exitPrice = trade.FinalExitPrice.Value;
            }
            else
            {
                exitPrice = trade.EntryPrice;
            }

            return exitPrice;
        }
    }
}
